### ISSUE TITLE
openpilot DM model compile in python 3.12 fails with recursion depth exceeded

### ISSUE THREAD

**Issue by haraschax** [tinygrad/tinygrad#8167]
Running 
```
GPU=1 python3 ./examples/openpilot/compile3.py https://github.com/commaai/openpilot/raw/13d4b5f8bb79c06811f4bee2c3542aca65292aac/selfdrive/modeld/models/dmonitoring_model.onnx
```

In python 3.12 produces:

```
(openpilot) batman@workstation-harald:~/xx/openpilot/tinygrad_repo$ GPU=1 python3 ./examples/openpilot/compile3.py https://github.com/commaai/openpilot/raw/13d4b5f8bb79c06811f4bee2c3542aca65292aac/selfdrive/modeld/models/dmonitoring_model.onnx
loaded model
created tensors
run 0
Traceback (most recent call last):
  File "/home/batman/xx/openpilot/tinygrad_repo/./examples/openpilot/compile3.py", line 122, in <module>
    test_val = compile(onnx_file) if not getenv("RUN") else None
               ^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/./examples/openpilot/compile3.py", line 44, in compile
    ret = run_onnx_jit(**inputs).numpy()
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/engine/jit.py", line 235, in __call__
    if len(params:=get_parameters(ret)): Tensor.realize(params[0], *params[1:])
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/tensor.py", line 3840, in _wrapper
    ret = fn(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/tensor.py", line 227, in realize
    run_schedule(*self.schedule_with_vars(*lst), do_update_stats=do_update_stats)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/tensor.py", line 3815, in _wrapper
    if _METADATA.get() is not None: return fn(*args, **kwargs)
                                           ^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/tensor.py", line 216, in schedule_with_vars
    schedule, var_vals = create_schedule_with_vars(flatten([x.lazydata.lbs for x in (self,)+lst]))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/ops.py", line 710, in __wrapper
    try: ret = func(self, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/engine/schedule.py", line 439, in create_schedule_with_vars
    big_graph = graph_rewrite(big_graph, ops_folding+do_realize, ctx)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/ops.py", line 783, in graph_rewrite
    return RewriteContext(pm, ctx).bottom_up_rewrite(sink) if bottom_up else RewriteContext(pm, ctx).rewrite(sink)
                                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/ops.py", line 768, in rewrite
    new_src = tuple(map(self.rewrite, n.src))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/ops.py", line 768, in rewrite
    new_src = tuple(map(self.rewrite, n.src))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad_repo/tinygrad/ops.py", line 768, in rewrite
    new_src = tuple(map(self.rewrite, n.src))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 744 more times]
RecursionError: maximum recursion depth exceeded
```

**Comment by chenyuxyz** [comment]
which subversion are you using? works on 3.12.8 on my mac

```
➜  tinygrad git:(master) ✗ python3.12 --version
Python 3.12.8
➜  tinygrad git:(master) ✗ PYTHONPATH=. GPU=1 python3.12 ./examples/openpilot/compile3.py https://github.com/commaai/openpilot/raw/13d4b5f8bb79c06811f4bee2c3542aca65292aac/selfdrive/modeld/models/dmonitoring_model.onnx
loaded model
created tensors
run 0
scheduled 570 kernels
memory reduced from 17.40 MB -> 15.36 MB, 310 -> 274 bufs
...
...
enqueue   1.15 ms -- total run   3.47 ms
enqueue   1.10 ms -- total run   3.32 ms
enqueue   1.22 ms -- total run   3.66 ms
enqueue   1.12 ms -- total run   3.40 ms
enqueue   1.13 ms -- total run   3.29 ms
enqueue   1.19 ms -- total run   3.60 ms
<Tensor <UOp GPU (1, 596) float (<Ops.VIEW: 13>, <buf real:True device:GPU size:596 dtype:dtypes.float offset:0>)> on GPU with grad None> (1, 596) float32
**** test done ****
```

**Comment by haraschax** [comment]
This is with 3.12.0 and ubuntu 20.04
```
openpilot) batman@workstation-harald:~/openpilot$ python
Cannot read termcap database;
using dumb terminal settings.
Python 3.12.0 (main, Oct  3 2023, 01:27:23) [Clang 17.0.1 ] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
KeyboardInterrupt
>>> ^D
(openpilot) batman@workstation-harald:~/openpilot$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 24.04.1 LTS
Release:	24.04
Codename:	noble
(openpilot) batman@workstation-harald:~/openpilot$ 
```

**Comment by haraschax** [comment]
Confirmed this is fixed on python 3.12.3.

**Comment by FeepingCreature** [comment]
$ python3 --version
Python 3.12.3
$ python3 sdxl.py --weights ~/sd_xl_base_1.0.safetensors 
Traceback (most recent call last):
  File "/home/feep/tinygrad/tinygrad/ops.py", line 897, in rewrite
    new_src = tuple(map(self.rewrite, n.src))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/feep/tinygrad/tinygrad/ops.py", line 897, in rewrite
    new_src = tuple(map(self.rewrite, n.src))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/feep/tinygrad/tinygrad/ops.py", line 897, in rewrite
    new_src = tuple(map(self.rewrite, n.src))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 997 more times]
RecursionError: maximum recursion depth exceeded
```
